// wechaty-puppet-padlocal/dist/cjs/src/padlocal/schema-mapper/message.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.padLocalMessageToWechaty = void 0;
const PUPPET = __importStar(require("wechaty-puppet"));
const mod_js_1 = require("./message/mod.js");
async function padLocalMessageToWechaty(puppet, padLocalMessage) {
    // set default value for MessagePayloadBase, other fields will be fulfilled or updated var MessageParers
    const ret = {
        id: padLocalMessage.id,
        talkerId: padLocalMessage.fromusername,
        text: padLocalMessage.content,
        timestamp: padLocalMessage.createtime,
        type: PUPPET.types.Message.Unknown,
    };
    await (0, mod_js_1.executeMessageParsers)(puppet, padLocalMessage, ret);
    // validate the return value
    if (!(ret.roomId || ret.listenerId)) {
        throw new Error("neither roomId nor listenerId");
    }
    return ret;
}
exports.padLocalMessageToWechaty = padLocalMessageToWechaty;
