// wechaty-puppet-padlocal/dist/cjs/src/padlocal/schema-mapper/room.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.chatRoomMemberToContact = exports.padLocalRoomMemberToWechaty = exports.padLocalRoomToWechaty = void 0;
const padlocal_pb_js_1 = __importDefault(require("padlocal-client-ts/dist/proto/padlocal_pb.js"));
function padLocalRoomToWechaty(contact) {
    return {
        adminIdList: [],
        avatar: contact.avatar,
        id: contact.username,
        memberIdList: contact.chatroommemberList.map((member) => member.username),
        ownerId: contact.chatroomownerusername,
        topic: contact.nickname,
    };
}
exports.padLocalRoomToWechaty = padLocalRoomToWechaty;
function padLocalRoomMemberToWechaty(chatRoomMember) {
    return {
        avatar: chatRoomMember.avatar,
        id: chatRoomMember.username,
        inviterId: chatRoomMember.inviterusername,
        name: chatRoomMember.nickname,
        roomAlias: chatRoomMember.displayname,
    };
}
exports.padLocalRoomMemberToWechaty = padLocalRoomMemberToWechaty;
function chatRoomMemberToContact(chatRoomMember) {
    return new padlocal_pb_js_1.default.Contact()
        .setUsername(chatRoomMember.getUsername())
        .setNickname(chatRoomMember.getNickname())
        .setAvatar(chatRoomMember.getAvatar())
        .setStranger(true);
}
exports.chatRoomMemberToContact = chatRoomMemberToContact;
