// wechaty-plugin-contrib/dist/src/mod.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.talkers = exports.mappers = exports.matchers = exports.finders = void 0;
const finders = __importStar(require("./finders/mod"));
exports.finders = finders;
const matchers = __importStar(require("./matchers/mod"));
exports.matchers = matchers;
const talkers = __importStar(require("./talkers/mod"));
exports.talkers = talkers;
const mappers = __importStar(require("./mappers/mod"));
exports.mappers = mappers;
var config_1 = require("./config");
Object.defineProperty(exports, "VERSION", { enumerable: true, get: function () { return config_1.VERSION; } });
var validate_plugin_1 = require("./validate-plugin");
Object.defineProperty(exports, "validatePlugin", { enumerable: true, get: function () { return validate_plugin_1.validatePlugin; } });
var mod_1 = require("./contrib/ding-dong/mod");
Object.defineProperty(exports, "DingDong", { enumerable: true, get: function () { return mod_1.DingDong; } });
var mod_2 = require("./contrib/heartbeat/mod");
Object.defineProperty(exports, "Heartbeat", { enumerable: true, get: function () { return mod_2.Heartbeat; } });
var mod_3 = require("./contrib/room-connector/mod");
Object.defineProperty(exports, "OneToManyRoomConnector", { enumerable: true, get: function () { return mod_3.OneToManyRoomConnector; } });
Object.defineProperty(exports, "ManyToOneRoomConnector", { enumerable: true, get: function () { return mod_3.ManyToOneRoomConnector; } });
Object.defineProperty(exports, "ManyToManyRoomConnector", { enumerable: true, get: function () { return mod_3.ManyToManyRoomConnector; } });
var qr_code_terminal_1 = require("./contrib/qr-code-terminal");
Object.defineProperty(exports, "QRCodeTerminal", { enumerable: true, get: function () { return qr_code_terminal_1.QRCodeTerminal; } });
var event_logger_1 = require("./contrib/event-logger");
Object.defineProperty(exports, "EventLogger", { enumerable: true, get: function () { return event_logger_1.EventLogger; } });
var chatops_1 = require("./contrib/chatops");
Object.defineProperty(exports, "ChatOps", { enumerable: true, get: function () { return chatops_1.ChatOps; } });
var friendship_accepter_1 = require("./contrib/friendship-accepter");
Object.defineProperty(exports, "FriendshipAccepter", { enumerable: true, get: function () { return friendship_accepter_1.FriendshipAccepter; } });
var room_inviter_1 = require("./contrib/room-inviter");
Object.defineProperty(exports, "RoomInviter", { enumerable: true, get: function () { return room_inviter_1.RoomInviter; } });
var event_hot_handler_1 = require("./contrib/event-hot-handler");
Object.defineProperty(exports, "EventHotHandler", { enumerable: true, get: function () { return event_hot_handler_1.EventHotHandler; } });
var room_invitation_accepter_1 = require("./contrib/room-invitation-accepter");
Object.defineProperty(exports, "RoomInvitationAccepter", { enumerable: true, get: function () { return room_invitation_accepter_1.RoomInvitationAccepter; } });
var message_awaiter_1 = require("./contrib/message-awaiter");
Object.defineProperty(exports, "MessageAwaiter", { enumerable: true, get: function () { return message_awaiter_1.MessageAwaiter; } });
